File: src/main/java/io/github/takenoha/towerdefense/tactical/TacticalTargetContext.java
package io.github.takenoha.towerdefense.tactical;

/** Read-only combat facts used when evaluating conditional tactical effects. */
public record TacticalTargetContext(
        double targetHealthFraction,
        double coreHealthFraction,
        boolean boss,
        boolean slowed,
        boolean burning) {
    public TacticalTargetContext {
        requireFraction(targetHealthFraction, "targetHealthFraction");
        requireFraction(coreHealthFraction, "coreHealthFraction");
    }

    public static TacticalTargetContext neutral() {
        return new TacticalTargetContext(1.0d, 1.0d, false, false, false);
    }

    public boolean targetHasHighHealth() {
        return targetHealthFraction >= 0.75d;
    }

    public boolean targetHasLowHealth() {
        return targetHealthFraction <= 0.30d;
    }

    public boolean coreBelowHalf() {
        return coreHealthFraction < 0.50d;
    }

    public boolean coreBelowThirtyPercent() {
        return coreHealthFraction < 0.30d;
    }

    private static void requireFraction(double value, String name) {
        if (!Double.isFinite(value) || value < 0.0d || value > 1.0d) {
            throw new IllegalArgumentException(name + " must be between 0 and 1");
        }
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/tactical/TacticalTargetContext.kt
package io.github.takenoha.towerdefense.tactical

import java.util.Objects
import kotlin.jvm.JvmRecord

/** Read-only combat facts used when evaluating conditional tactical effects. */
@JvmRecord
data class TacticalTargetContext(
    val targetHealthFraction: Double,
    val coreHealthFraction: Double,
    val boss: Boolean,
    val slowed: Boolean,
    val burning: Boolean,
) {
    init {
        requireFraction(targetHealthFraction, "targetHealthFraction")
        requireFraction(coreHealthFraction, "coreHealthFraction")
    }

    fun targetHasHighHealth(): Boolean = targetHealthFraction >= 0.75

    fun targetHasLowHealth(): Boolean = targetHealthFraction <= 0.30

    fun coreBelowHalf(): Boolean = coreHealthFraction < 0.50

    fun coreBelowThirtyPercent(): Boolean = coreHealthFraction < 0.30

    companion object {
        @JvmStatic
        fun neutral(): TacticalTargetContext =
            TacticalTargetContext(1.0, 1.0, false, false, false)

        private fun requireFraction(value: Double, name: String) {
            Objects.requireNonNull(name, "name")
            if (!value.isFinite() || value < 0.0 || value > 1.0) {
                throw IllegalArgumentException("$name must be between 0 and 1")
            }
        }
    }
}
